lib/sqlalchemy/sql/elements.py

```diff
@@ -2499,21 +2543,25 @@ def columns(self, *cols, **types):
         )
 
     @property
-    def type(self):
+    def type(self) -> NullType:
         return type_api.NULLTYPE
 
     @property
-    def comparator(self):
-        return self.type.comparator_factory(self)
+    def comparator(self) -> NullType.Comparator[Any]:
+        return self.type.comparator_factory(self)  # type: ignore [arg-type]
 
-    def self_group(self, against=None):
+    def self_group(
+        self, against: Optional[OperatorType] = None
+    ) -> Union[Grouping[TextClause], TextClause]:
```

Comment:
`self_group()` has a lot of overrides with potentially conflicting signatures. It is possible to narrow some types down with `@overload`s matching `self` and return types, but seems overkill, so most of the time I annotated the return with superclass  `ColumnElement` or `ClauseElement`. Here it was possible to narrow down types without tripping mypy.

Reply:
Inside `ColumnElement` there is an implementation of `self_group()` with two `@overload`s that seem redundant - but mypy is happy. Would be nice to ask the author about them, because I cannot understand what's going on there

Reply:
I think those overloads are needed to cover both cases where we return just `self` `(L1402)` without loosing information about the contained type and where we return `self` wrapped into other classes which conceal `self`'s contained type `(L1398)` and `(L1400)`. Not sure if it's actually the case but it's the best idea I can come up with off the top of my head.

https://github.com/GodelTech/sqlalchemy/blob/c6893624a6e63b4e532ce5f8b6f5e86e29abab44/lib/sqlalchemy/sql/elements.py#L1379-L1402

Reply:
That's a fair point and might indeed be what the author intended, but I fail to see how **exactly** is this going to work technically. [Normally](https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-calls-to-overloads), Mypy would choose the best matching overload, and if there are several candidates, it chooses whichever one comes first in the declaration. However, if there is `Any` among the candiates - Mypy just infers `Any`. So here, when Mypy resolves `ColumnElement`s type argument, why would it ever pick `ColumnElement[_T]`?